File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/F._Расстояние_между_вершинами/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(V + E)
    Memory Complexity O(V + E)
    """
    n, m = map(int, sys.stdin.readline().rstrip().split())
    adj_list: dict[int, list[int]] = dict()
    for _ in range(m):
        u, v = map(int, sys.stdin.readline().rstrip().split())
        if u not in adj_list: adj_list[u] = []
        adj_list[u].append(v)
        if v not in adj_list: adj_list[v] = []
        adj_list[v].append(u)
    source, destination = map(int, sys.stdin.readline().rstrip().split())
    edges: int = 0
    seen: list[bool] = [False] * (n + 1)
    cur_nodes: list[int] = [source]
    next_nodes: list[int] = []
    while cur_nodes:
        for node in cur_nodes:
            if node == destination:
                sys.stdout.write(f'{edges}\n')
                return
            for neighbor in adj_list.get(node, []):
                if not seen[neighbor]:
                    seen[neighbor] = True
                    next_nodes.append(neighbor)
        edges += 1
        cur_nodes = next_nodes
        next_nodes = []
    sys.stdout.write(f'-1\n')


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/F._Расстояние_между_вершинами/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_map>


void solution() {
    /*
    Time Complexity O(V + E)
    Memory Complexity O(V + E)
    */
    int n, m;
    std::cin >> n >> m;
    std::unordered_map<int, std::vector<int>> adjList;
    for (int i = 0; i < m; ++i) {
        int u, v;
        std::cin >> u >> v;
        adjList[u].push_back(v);
        adjList[v].push_back(u);
    }
    int source, destination;
    std::cin >> source >> destination;
    std::vector<bool> seen (n + 1, false);
    std::vector<int> curNodes = {source};
    seen[source] = true;
    std::vector<int> nextNodes;
    int edges = 0;
    while (!curNodes.empty()) {
        for (int& node : curNodes) {
            if (node == destination) {
                std::cout << edges << "\n";
                return;
            }
            for (int& neighbor : adjList[node]) {
                if (!seen[neighbor]) {
                    nextNodes.push_back(neighbor);
                    seen[neighbor] = true;
                }
            }
        }
        ++edges;
        curNodes = nextNodes;
        nextNodes = {};
    }
    std::cout << "-1\n";
}


int main() {
    solution();
}